src/murmur/Messages.cpp

```diff
@@ -195,34 +195,48 @@ void Server::msgAuthenticate(ServerUser *uSource, MumbleProto::Authenticate &msg
 
 	bool ok     = false;
 	bool nameok = validateUserName(uSource->qsName);
-	QString pw  = u8(msg.password());
-
-	// Fetch ID and stored username.
-	// Since this may call DBus, which may recall our dbus messages, this function needs
-	// to support re-entrancy, and also to support the fact that sessions may go away.
-	int id = authenticate(uSource->qsName, pw, uSource->uiSession, uSource->qslEmail, uSource->qsHash,
-						  uSource->bVerified, uSource->peerCertificateChain());
-
-	uSource->iId = id >= 0 ? id : -1;
 
 	QString reason;
 	MumbleProto::Reject_RejectType rtType = MumbleProto::Reject_RejectType_None;
+	int id                                = -2;
+
+	if (!bForceUsernameCertSubjectEquality
+		|| (bCertRequired
+			&& uSource->qsName
+				   == uSource->peerCertificateChain()
+						  .first() // First in chain is the users certificate
+						  .subjectInfo(QSslCertificate::SubjectInfo::CommonName)
+						  .join(""))) {
+		QString pw = u8(msg.password());
+
+		// Fetch ID and stored username.
+		// Since this may call DBus, which may recall our dbus messages, this function needs
+		// to support re-entrancy, and also to support the fact that sessions may go away.
```

Comment:
```suggestion
		// As sessions may disappear anytime, this function needs to be re-entrant.
		// It's also to prevent issues with DBus messages.
```